### ISSUE TITLE
BE crash when  compaction

### ISSUE THREAD

**Issue by tiannan-sr** [StarRocks/starrocks#2717]
### Steps to reproduce the behavior (Required)
BE's parameter: max_compaction_concurrency = 0,
create tables and import data.
set max_compaction_concurrency = -1 and restart BE, BE crash.

### Expected behavior (Required)
compaction task succeed.

### Real behavior (Required)
BE crash when compaction:
```
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
*** Aborted at 1641645916 (unix time) try "date -d @1641645916" if you are using GNU date ***
src/central_freelist.cc:333] tcmalloc: allocation failed 8192
src/central_freelist.cc:333] tcmalloc: allocation failed 8192
src/central_freelist.cc:333] tcmalloc: allocation failed 8192
src/central_freelist.cc:333] tcmalloc: allocation failed 8192
src/central_freelist.cc:333] tcmalloc: allocation failed 8192
src/central_freelist.cc:333] tcmalloc: allocation failed 8192
terminate called recursively
terminate called recursively
terminate called recursively
terminate called recursively
terminate called recursively
terminate called recursively
PC: @     0x7f0bf3dca387 __GI_raise
*** SIGABRT (@0x3e80000339b) received by PID 13211 (TID 0x7f0b73100700) from PID 13211; stack trace: ***
    @          0x342f222 google::(anonymous namespace)::FailureSignalHandler()
    @     0x7f0bf4a95630 (unknown)
    @     0x7f0bf3dca387 __GI_raise
    @     0x7f0bf3dcba78 __GI_abort
    @          0x15a4d83 _ZN9__gnu_cxx27__verbose_terminate_handlerEv.cold
    @          0x4de3e66 __cxxabiv1::__terminate()
    @          0x4de3ed1 std::terminate()
    @          0x4de4025 __cxa_throw
    @          0x15a4c8a _Znwm.cold
    @          0x180c20d std::__cxx11::basic_string<>::_M_mutate()
    @          0x180c3ca std::__cxx11::basic_string<>::resize()
    @          0x1805d99 starrocks::Segment::parse_segment_footer()
src/central_freelist.cc:333] tcmalloc: allocation failed 40960
src/central_freelist.cc:333] tcmalloc: allocation failed 40960
src/central_freelist.cc:333] tcmalloc: allocation failed 40960
src/central_freelist.cc:333] tcmalloc: allocation failed 8192
terminate called recursively
    @          0x18069e3 starrocks::Segment::_open()
    @          0x1806e89 starrocks::Segment::open()
    @          0x1bcc0b4 starrocks::BetaRowset::do_load()
    @          0x1bc1afc starrocks::Rowset::load()
    @          0x1bcc719 starrocks::BetaRowset::get_segment_iterators()
    @          0x189479e starrocks::vectorized::TabletReader::get_segment_iterators()
    @          0x18ac82f starrocks::vectorized::Compaction::_get_segment_iterator_num()
    @          0x18b1e0d starrocks::vectorized::Compaction::do_compaction_impl()
    @          0x18b389b starrocks::vectorized::Compaction::do_compaction()
    @          0x18aa8d7 starrocks::vectorized::CumulativeCompaction::compact()
    @          0x1737129 starrocks::StorageEngine::_perform_cumulative_compaction()
    @          0x17247cf starrocks::StorageEngine::_cumulative_compaction_thread_callback()
    @          0x4e5e000 execute_native_thread_routine
    @     0x7f0bf4a8dea5 start_thread
    @     0x7f0bf3e929fd __clone
    @                0x0 (unknown)
```

BE's mem_metrics before crash:
```
Sat Jan 8 20:45:15 CST 2022
# TYPE starrocks_be_compaction_mem_bytes gauge
starrocks_be_compaction_mem_bytes 43195240032
# TYPE starrocks_be_load_mem_bytes gauge
starrocks_be_load_mem_bytes 25381664
# TYPE starrocks_be_process_mem_bytes gauge
starrocks_be_process_mem_bytes 51934051480
# TYPE starrocks_be_query_mem_bytes gauge
starrocks_be_query_mem_bytes 0
# TYPE starrocks_be_schema_change_mem_bytes gauge
starrocks_be_schema_change_mem_bytes 0
# TYPE starrocks_be_tablet_meta_mem_bytes gauge
starrocks_be_tablet_meta_mem_bytes 32275666892
# TYPE starrocks_be_tcmalloc_bytes_in_use gauge
starrocks_be_tcmalloc_bytes_in_use 51950507008
# TYPE starrocks_be_update_mem_bytes gauge
starrocks_be_update_mem_bytes 0
Sat Jan 8 20:45:16 CST 2022
# TYPE starrocks_be_compaction_mem_bytes gauge
starrocks_be_compaction_mem_bytes 44145089696
# TYPE starrocks_be_load_mem_bytes gauge
starrocks_be_load_mem_bytes 25381664
# TYPE starrocks_be_process_mem_bytes gauge
starrocks_be_process_mem_bytes 52883901144
# TYPE starrocks_be_query_mem_bytes gauge
starrocks_be_query_mem_bytes 0
# TYPE starrocks_be_schema_change_mem_bytes gauge
starrocks_be_schema_change_mem_bytes 0
# TYPE starrocks_be_tablet_meta_mem_bytes gauge
starrocks_be_tablet_meta_mem_bytes 32835851788
# TYPE starrocks_be_tcmalloc_bytes_in_use gauge
starrocks_be_tcmalloc_bytes_in_use 52898202384
# TYPE starrocks_be_update_mem_bytes gauge
starrocks_be_update_mem_bytes 0
Sat Jan 8 20:45:17 CST 2022
Sat Jan 8 20:45:23 CST 2022
Sat Jan 8 20:45:24 CST 2022
```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
mysql> select current_version();
+--------------------------+
| current_version()        |
+--------------------------+
| STABILITY_MASTER 3aff094 |
+--------------------------+
1 row in set (0.35 sec)
```

**Comment by wyb** [comment]
In the scenario of 10 thousand columns schema, tablet meta takes up a lot of memory. (32g/50g from the memory metrics)

**Comment by xiaoyong-z** [comment]
The problem cannot be reproduced, it seems to have been fixed by the other memory optimization pr in 2.3.
